// @chakra-ui/card/dist/index.esm.js
// ../../../react-shim.js
import React from "react";

// ../../utilities/shared-utils/src/index.ts
var cx = (...classNames) => classNames.filter(Boolean).join(" ");

// src/card.tsx
import {
  forwardRef,
  chakra,
  omitThemingProps,
  useMultiStyleConfig
} from "@chakra-ui/system";

// src/card-context.tsx
import { createStylesContext } from "@chakra-ui/system";
var [CardStylesProvider, useCardStyles] = createStylesContext("Card");

// src/card.tsx
var Card = forwardRef(function Card2(props, ref) {
  const {
    className,
    children,
    direction = "column",
    justify,
    align,
    ...rest
  } = omitThemingProps(props);
  const styles = useMultiStyleConfig("Card", props);
  return /* @__PURE__ */ React.createElement(chakra.div, {
    ref,
    className: cx("chakra-card", className),
    __css: {
      display: "flex",
      flexDirection: direction,
      justifyContent: justify,
      alignItems: align,
      position: "relative",
      minWidth: 0,
      wordWrap: "break-word",
      ...styles.container
    },
    ...rest
  }, /* @__PURE__ */ React.createElement(CardStylesProvider, {
    value: styles
  }, children));
});

// src/card-body.tsx
import { forwardRef as forwardRef2, chakra as chakra2 } from "@chakra-ui/system";
var CardBody = forwardRef2(function CardBody2(props, ref) {
  const { className, ...rest } = props;
  const styles = useCardStyles();
  return /* @__PURE__ */ React.createElement(chakra2.div, {
    ref,
    className: cx("chakra-card__body", className),
    __css: styles.body,
    ...rest
  });
});

// src/card-footer.tsx
import {
  forwardRef as forwardRef3,
  chakra as chakra3
} from "@chakra-ui/system";
var CardFooter = forwardRef3(function CardFooter2(props, ref) {
  const { className, justify, ...rest } = props;
  const styles = useCardStyles();
  return /* @__PURE__ */ React.createElement(chakra3.div, {
    ref,
    className: cx("chakra-card__footer", className),
    __css: {
      ...styles.footer,
      display: "flex",
      justifyContent: justify
    },
    ...rest
  });
});

// src/card-header.tsx
import { forwardRef as forwardRef4, chakra as chakra4 } from "@chakra-ui/system";
var CardHeader = forwardRef4(function CardHeader2(props, ref) {
  const { className, ...rest } = props;
  const styles = useCardStyles();
  return /* @__PURE__ */ React.createElement(chakra4.div, {
    ref,
    className: cx("chakra-card__header", className),
    __css: styles.header,
    ...rest
  });
});
export {
  Card,
  CardBody,
  CardFooter,
  CardHeader,
  useCardStyles
};
